File: FAB.FS/FAB.Model.FSharp/GameBoardFunctions.fs
module GameBoardFunctions
open FAB.Types
open System
open TechnicalServices

let readSquare (board:GameBoard, loc:Location)=
    if board.Ships |> Seq.exists (fun (ship: Ship) -> ShipFunctions.isHitInLocation(ship, loc)) then
        SquareValues.Hit
    else if board.Misses |> Seq.contains loc then
        SquareValues.Miss
    else 
        SquareValues.Empty

let allSunk ships  = not (ships |> Seq.exists(fun (ship : Ship)-> not(ShipFunctions.isSunk ship)))

let rec getLocations (start: Location) orient (numberToAdd: int)  =
    if numberToAdd = 0 then
        List.empty
    else
        match orient with
        |Orientations.Horizontal -> start :: getLocations (start.Add 1 0) orient (numberToAdd - 1) 
        |_ -> start :: getLocations (start.Add 0 1) orient (numberToAdd - 1)   

let fitsWithinBoundaries  boardSize  (ship: Ship) =
    match ship.Orientation with
     |Orientations.Horizontal -> ship.Location.Col + ship.Size <= boardSize
     |_ ->  ship.Location.Row + ship.Size <= boardSize

let isValidPosition boardSize existingShips ship =
    if not(fitsWithinBoundaries boardSize ship) then
        false
    else
        let anyShipOccupiesLocation ships loc =
            ships |> Seq.exists( fun (ship: Ship) -> ShipFunctions.occupies(ship, loc))
        let locs= getLocations ship.Location ship.Orientation ship.Size
        not (locs |> Seq.exists( fun loc -> anyShipOccupiesLocation existingShips loc))
       
let getRandomPosition boardSize  (random: Random) =
    let result1 = RandomNumbers.Next(random, 0, boardSize)
    let col = result1.Number
    let result2 = RandomNumbers.Next(result1.NewGenerator, 0, boardSize)
    let row = result2.Number
    let result3 = RandomNumbers.Next(result2.NewGenerator, 0, 2)
    let orientation : Orientations = LanguagePrimitives.EnumOfValue result3.Number
    let loc = new Location(col, row)
    (loc, orientation, result3.NewGenerator)

let rec setValidRandomPosition boardSize shipsAlreadyPlaced  (shipToBePlaced:Ship) random = 
    let loc,orient,newRandom = getRandomPosition boardSize  random
    let placedShip = new Ship(shipToBePlaced.Name, shipToBePlaced.Size, loc, orient)
    if isValidPosition boardSize shipsAlreadyPlaced placedShip then
        (placedShip, newRandom)
    else
        setValidRandomPosition boardSize  shipsAlreadyPlaced  shipToBePlaced  newRandom

let rec locateShipsRandomly boardSize (shipsToBePlaced: seq<Ship>) shipsAlreadyPlaced random =
    if shipsToBePlaced |> Seq.isEmpty then
        List.empty
    else 
        let thisShip = shipsToBePlaced |> Seq.head;
        let message = "Computer placing the " + thisShip.Name + "\n";
        let placedShip, newRandom = setValidRandomPosition boardSize  shipsAlreadyPlaced  thisShip  random
        let shipPlacement = (placedShip, message)
        let newshipsAlreadyPlaced = placedShip :: shipsAlreadyPlaced
        let newShipsToBePlaced = shipsToBePlaced |> Seq.skip 1 
        shipPlacement :: locateShipsRandomly boardSize newShipsToBePlaced newshipsAlreadyPlaced newRandom
    
let createBoardWithShipsPlacedRandomly boardSize shipsToBePlaced random =
    let shipPlacements = locateShipsRandomly boardSize shipsToBePlaced List.Empty random
    let newShips = shipPlacements |> Seq.map(fun sp -> fst sp)
    let messages = shipPlacements |> Seq.map(fun sp -> snd sp)
    let aggregateMsg = messages |> Seq.fold (+) ""
    new GameBoard(boardSize, newShips, aggregateMsg, List.Empty); 

let checkSquareAndRecordOutcome (board: GameBoard) loc aggregateMessages =
    let results = board.Ships |> Seq.map(fun (s: Ship) -> ShipFunctions.fireAt(s, loc))
    let newShips = results |> Seq.map(fun r -> match r with (ship,_,_) -> ship)
    let hits = results |> Seq.map(fun r -> match r with (_,hit,_) -> hit)
    let hit = hits |> Seq.fold (||) false
    let messages = results |> Seq.map(fun r -> match r with (_,_,msg) -> msg)
    let foldedMessages = messages |> Seq.fold (+) ""
    let aggregatedMessages = if aggregateMessages then board.Messages + foldedMessages else foldedMessages
    let misses = board.Misses;
    if hit then
        if allSunk newShips then
            let allSunk = foldedMessages + "All ships sunk!"
            new GameBoard(board.Size, newShips, allSunk, misses);
        else
             new GameBoard(board.Size, newShips, aggregatedMessages, misses);
    else
        let newMisses = loc :: Seq.toList(board.Misses)
        let sorry = aggregatedMessages + "Sorry, (" + loc.Col.ToString() + "," + loc.Row.ToString() + ") is a miss.";
        new GameBoard(board.Size, newShips, sorry, newMisses);

    
let rec checkSquaresAndRecordOutcome board (locs : List<Location>) =
    let boardFromHead = checkSquareAndRecordOutcome board locs.Head true
    if locs.Length = 1 then
        boardFromHead
    else
        checkSquaresAndRecordOutcome boardFromHead locs.Tail;

File: FAB/FAB.Model.CSharp/GameboardFunctions.cs
using Quadrivia.FunctionalLibrary;
using System;
using System.Collections.Immutable;
using System.Linq;

namespace Quadrivia.FAB
{
    public static class GameBoardFunctions
    {
        private const string newLine = "\n";

        public static bool allSunk(FList<Ship> ships)
        {
            return !FList.Any(ship => !ship.isSunk(), ships);
        }

        public static GameBoard checkSquareAndRecordOutcome(this GameBoard board, Location loc, bool aggregate = false)
        {
            return HitSomething(board, loc) ?
                allSunk(ShipsAfterFiring(board, loc)) ?
                     new GameBoard(board.Size, ShipsAfterFiring(board, loc), MessagesAfterFiring(board, loc) + "All ships sunk!", board.Misses)
                     : new GameBoard(board.Size, ShipsAfterFiring(board, loc), AggregatedMessages(board, aggregate, MessagesAfterFiring(board, loc)), board.Misses)
                : new GameBoard(board.Size, ShipsAfterFiring(board, loc), AddMissMsg(board, loc, aggregate, MessagesAfterFiring(board, loc)), board.Misses.Add(loc));
        }

        private static bool HitSomething(GameBoard board, Location loc)
        {
            return FList.FoldL((a, b) => a | b, false, FList.Map(s => s.fireAt(loc).Item2, board.Ships));
        }
        //TODO: Check use of FoldL vs FoldR here and above
        private static string MessagesAfterFiring(GameBoard board, Location loc)
        {
            return FList.FoldL((a, b) => a + b, "",FList.Map(s => s.fireAt(loc).Item3, board.Ships));
        }

        private static FList<Ship> ShipsAfterFiring(GameBoard board, Location loc)
        {
            return FList.Map(s => s.fireAt(loc).Item1, board.Ships);
        }

        private static string AddMissMsg(GameBoard board, Location loc, bool aggregate, string newMessages)
        {
            return AggregatedMessages(board, aggregate, newMessages) + "Sorry, (" + loc.Col + "," + loc.Row + ") is a miss.";
        }

        private static string AggregatedMessages(GameBoard board, bool aggregateMessages, string newMessages)
        {
            return aggregateMessages ? board.Messages + newMessages : newMessages;
        }

        public static GameBoard checkSquaresAndRecordOutcome(this GameBoard board, FList<Location> locs)
        {
            return FList.Length(locs) == 1 ?
                checkSquareAndRecordOutcome(board, FList.Head(locs), true)
                : checkSquaresAndRecordOutcome(checkSquareAndRecordOutcome(board, FList.Head(locs), true), FList.RemoveFirst(FList.Head(locs), locs));
        }

        public static bool isValidPosition(int boardSize, FList<Ship> existingShips, Ship shipToBePlaced)
        {
            return !shipWouldFitWithinBoard(boardSize, shipToBePlaced) ?
                 false
                 : !FList.Any(s => s.intersects(shipToBePlaced), existingShips);
        }

        public static FList<Location> locationsThatShipWouldOccupy(Location loc, Orientations orient, int locsToAdd)
        {
            return locsToAdd == 0 ?
                 FList.Empty<Location>()
                 : orient == Orientations.Horizontal ?
                         locationsThatShipWouldOccupy(loc.Add(1, 0), orient, locsToAdd - 1)
                         : FList.Prepend(loc,locationsThatShipWouldOccupy(loc.Add(0, 1), orient, locsToAdd - 1));
        }

        public static bool shipWouldFitWithinBoard(int boardSize, Ship ship)
        {
            return (ship.Orientation == Orientations.Horizontal
                        && ship.Location.Col + ship.Size <= boardSize)
                    ||
                    (ship.Orientation == Orientations.Vertical
                        && ship.Location.Row + ship.Size <= boardSize);
        }

        public static bool contains(this GameBoard board, Location loc)
        {
            return loc.Col >= 0 && loc.Col < board.Size && loc.Row >= 0 && loc.Row < board.Size;
        }

        public static SquareValues readSquare(this GameBoard board, Location loc)
        {
            return FList.Any(s => s.isHitInLocation(loc), board.Ships) ?
                SquareValues.Hit
                : board.Misses.Contains(loc) ?
                    SquareValues.Miss
                    : SquareValues.Empty;
        }

        public static GameBoard createBoardWithShipsPlacedRandomly(int boardSize, FList<Ship> shipsToBePlaced, FRandom random)
        {
            return new GameBoard(
                boardSize,
                FList.Map(r => r.Item1,locateShipsRandomly(boardSize, shipsToBePlaced, FList.Empty<Ship>(), random)),
                FList.FoldL((r, s) => r + s, "", FList.Map(r => r.Item2,locateShipsRandomly(boardSize, shipsToBePlaced, FList.Empty<Ship>(), random))),
                ImmutableHashSet.Create<Location>()
            );
        }

        public static FList<Tuple<Ship, string>> locateShipsRandomly(int boardSize, FList<Ship> shipsToBePlaced, FList<Ship> shipsAlreadyPlaced, FRandom random)
        {
            return FList.Length(shipsToBePlaced)== 0 ?
                FList.Empty<Tuple<Ship, string>>()
                : FList.New(
                    Tuple.Create(
                        locateShipRandomly(boardSize, shipsAlreadyPlaced, FList.Head(shipsToBePlaced), random).Item1,
                        locateShipRandomly(boardSize, shipsAlreadyPlaced, FList.Head(shipsToBePlaced), random).Item2
                    ),
                    locateShipsRandomly(
                        boardSize,
                        FList.RemoveFirst(FList.Head(shipsToBePlaced), shipsToBePlaced),
                        FList.Prepend(
                            locateShipRandomly(boardSize, shipsAlreadyPlaced, FList.Head(shipsToBePlaced), random).Item1, shipsAlreadyPlaced),
                        locateShipRandomly(boardSize, shipsAlreadyPlaced, FList.Head(shipsToBePlaced), random).Item3
                    )
                );
        }

        public static Tuple<Ship, string, FRandom> locateShipRandomly(
            int boardSize, FList<Ship> shipsAlreadyLocated, Ship shipToBeLocated, FRandom random)
        {
            return Tuple.Create(
                shipToBeLocated.setPosition(
                    getValidRandomPosition(boardSize, shipsAlreadyLocated, shipToBeLocated, random).Item1,
                    getValidRandomPosition(boardSize, shipsAlreadyLocated, shipToBeLocated, random).Item2),
                "Computer placing the " + shipToBeLocated.Name + newLine,
                getValidRandomPosition(boardSize, shipsAlreadyLocated, shipToBeLocated, random).Item3);
        }

        public static Tuple<Location, Orientations, FRandom> getValidRandomPosition(int boardSize, FList<Ship> shipsAlreadyLocated, Ship shipToBeLocated, FRandom random)
        {
            return isValidPosition(
                    boardSize,
                    shipsAlreadyLocated,
                    new Ship(shipToBeLocated.Name, shipToBeLocated.Size, getRandomPosition(boardSize, random).Item1,
                    getRandomPosition(boardSize, random).Item2)
                ) ?
                getRandomPosition(boardSize, random)
                : getValidRandomPosition(
                    boardSize,
                    shipsAlreadyLocated,
                    shipToBeLocated,
                    getRandomPosition(boardSize, random).Item3
                  );
        }

        public static Tuple<Location, Orientations, FRandom> getRandomPosition(int boardSize, FRandom random)
        {
            return Tuple.Create(
                new Location(FRandom.Skip(0, random, 0, boardSize).Number, FRandom.Skip(1, random, 0, boardSize).Number),
                (Orientations)FRandom.Skip(2, random, 0, 2).Number,
                FRandom.Skip(2, random, 0, 2)); //Deliberately uses same as the last call -  this contains the next seed.
        }

        public static Location Add(this Location loc, int colInc, int rowInc)
        {
            return new Location(loc.Col + colInc, loc.Row + rowInc);
        }
    }
}